Src/DependencyCollector/Shared/Implementation/TelemetryDiagnosticSourceListener.cs

```diff
@@ -0,0 +1,186 @@
+namespace Microsoft.ApplicationInsights.DependencyCollector.Implementation
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using Microsoft.ApplicationInsights.DataContracts;
+    using Microsoft.ApplicationInsights.Extensibility;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation;
+    using Microsoft.ApplicationInsights.Web.Implementation;
+    using Microsoft.ApplicationInsights.Channel;
+
+    internal class TelemetryDiagnosticSourceListener : DiagnosticSourceListenerBase<HashSet<string>>
+    {
+        internal const string ActivityStartNameSuffix = ".Start";
+        internal const string ActivityStopNameSuffix = ".Stop";
+
+        private readonly HashSet<string> excludedDiagnosticSources 
+            = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+        private readonly Dictionary<string, HashSet<string>> excludedDiagnosticSourceActivities 
+            = new Dictionary<string, HashSet<string>>(StringComparer.OrdinalIgnoreCase);
+
+
+        public TelemetryDiagnosticSourceListener(TelemetryConfiguration configuration, ICollection<string> excludeDiagnosticSourceActivities) 
+            : base(configuration)
+        {
+            this.Client.Context.GetInternalContext().SdkVersion = SdkVersionUtils.GetSdkVersion("rdd" + RddSource.DiagnosticSourceListener + ":");
+            this.PrepareExclusionLists(excludeDiagnosticSourceActivities);
+        }
+
+        internal override bool IsSourceEnabled(DiagnosticListener value)
+        {
+            return !this.excludedDiagnosticSources.Contains(value.Name);
+        }
+
+        protected override HashSet<string> GetListenerContext(DiagnosticListener diagnosticListener)
+        {
+            HashSet<string> excludedActivities;
+            if (!this.excludedDiagnosticSourceActivities.TryGetValue(diagnosticListener.Name, out excludedActivities))
+            {
+                return null;
+            }
+
+            return excludedActivities;
+        }
+
+        internal override bool IsEventEnabled(string evnt, object input1, object input2, DiagnosticListener diagnosticListener, HashSet<string> context)
+        {
+            return !this.IsActivityExcluded(evnt, context)
+                && !evnt.EndsWith(ActivityStartNameSuffix, StringComparison.OrdinalIgnoreCase);
+        }
+
+        internal override bool ShouldHandleEvent(KeyValuePair<string, object> evnt, DiagnosticListener diagnosticListener, HashSet<string> context)
```

Comment:
what is the point in having IsEventEnabled and ShouldHandleEvent? it's pure overhead and I don't think there is a real case for ShouldHandleEvent. In any case, listener may just do this check in HandleEvent method

Reply:
In case of Activities - ```IsEventEnabled``` will return ```true``` for events with activity name and activity stop name, while ```ShouldHandleEvent``` will only return ```true``` for the stop event. 

Yes, ```ShouldHandleEvent``` could be merged into ```HandleEvent``` however I was considering a default implementation which would work for non-activity cases.

Reply:
so it means that every listener would sacrifice some of the performance on the hottest path (IsEnabled being called twice) to enable easier support for non-Activity listeners? I'd prefer to keep non-Activity listeners overriding IsEnabled and Handle however they want, but keeping most interesting, popular and heavy loaded scenarios faster

Reply:
I've already removed ```ShouldHandleEvent``` as suggested